### PR TITLE
feat: Expose Python C headers through the toolchain.

### PR DESCRIPTION
**Pull request by rickeylev**
This allows getting a build's `cc_library` of Python headers through
toolchain resolution instead of having to use the underlying toolchain's
repository `:python_headers` target directly.

Without this feature, it's not possible to reliably and correctly get
the C information about the runtime a build is going to use. Existing
solutions require carefully setting up repo names, external state,
and/or using specific build rules. In comparison, with this feature,
consumers are able to simply ask for the current headers via a helper
target or manually lookup the toolchain and pull the relevant
information; toolchain resolution handles finding the correct headers.

The basic way this works is by registering a second toolchain to carry
C/C++ related information; as such, it is named `py_cc_toolchain`. The
py cc toolchain has the same constraint settings as the regular py
toolchain; an expected invariant is that there is a 1:1 correspondence
between the two. This base functionality allows a consuming rule
implementation to use toolchain resolution to find the Python C
toolchain information.

Usually what downstream consumers need are the headers to feed into
another `cc_library` (or equivalent) target, so, rather than have every
project re-implement the same "lookup and forward cc_library info" logic,
this is provided by the `//python/cc:current_py_cc_headers` target.
Targets that need the headers can then depend on that target as if it
was a `cc_library` target.

Work towards https://github.com/bazelbuild/rules_python/issues/824

### CODE REVIEW

**Comment by rickeylev** [comment]
Ok, CI is happy now -- ready for review.

File: python/cc/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# Package for C/C++ specific functionality of the Python rules.
+
+load("@bazel_skylib//:bzl_library.bzl", "bzl_library")
+load("//python/private:current_py_cc_headers.bzl", "current_py_cc_headers")
+load("//python/private:util.bzl", "BZLMOD_ENABLED")
```
Comment (chrislovecnm) [inline python/cc/BUILD.bazel:5]: @rickeylev 👍 - We needed this.  We need to file an issue to refactor everywhere to use this.

File: python/cc/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# Package for C/C++ specific functionality of the Python rules.
+
+load("@bazel_skylib//:bzl_library.bzl", "bzl_library")
+load("//python/private:current_py_cc_headers.bzl", "current_py_cc_headers")
```
Comment (chrislovecnm) [inline python/cc/BUILD.bazel:4]: We need to file and issue to use `//` instead of `@` when using rules Python internally.

File: python/private/current_py_cc_headers.bzl [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+def _current_py_cc_headers_impl(ctx):
+    py_cc_toolchain = ctx.toolchains["//python/cc:toolchain_type"].py_cc_toolchain
+    return py_cc_toolchain.headers.providers_map.values()
+
+current_py_cc_headers = rule(
+    implementation = _current_py_cc_headers_impl,
+    toolchains = ["//python/cc:toolchain_type"],
+    provides = [CcInfo],
+    doc = """\
+Provides the currently active Python toolchain's C headers.
```
Comment (chrislovecnm) [inline python/private/current_py_cc_headers.bzl:26]: Does this provide all of the headers, including C++?

File: python/private/current_py_cc_headers.bzl [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+This is basically a wrapper around the underlying `cc_library()` for the
```
Comment (chrislovecnm) [inline python/private/current_py_cc_headers.bzl:28]: > This is basically

Remove the word "basically" please.

File: python/private/current_py_cc_headers.bzl [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+C headers for the consuming target's currently active Python toolchain.
+
+To use, simply depend on this target where you would have wanted the
```
Comment (chrislovecnm) [inline python/private/current_py_cc_headers.bzl:31]: Provide an example, please.

File: python/private/py_cc_toolchain_rule.bzl [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+"""Implementation of py_cc_toolchain rule."""
+
+load(":py_cc_toolchain_info.bzl", "PyCcToolchainInfo")
+
+def _py_cc_toolchain_impl(ctx):
+    py_cc_toolchain = PyCcToolchainInfo(
+        headers = struct(
+            providers_map = {
+                "CcInfo": ctx.attr.headers[CcInfo],
```
Comment (chrislovecnm) [inline python/private/py_cc_toolchain_rule.bzl:23]: Should we check to make sure the values in the map exist?

File: python/private/py_cc_toolchain_rule.bzl [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                "DefaultInfo": ctx.attr.headers[DefaultInfo],
+            },
+        ),
+        python_version = ctx.attr.python_version,
+    )
+    return [platform_common.ToolchainInfo(
+        py_cc_toolchain = py_cc_toolchain,
+    )]
+
+py_cc_toolchain = rule(
```
Comment (chrislovecnm) [inline python/private/py_cc_toolchain_rule.bzl:33]: Please add `doc =` to the rule itself.

File: python/private/py_cc_toolchain_rule.bzl [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+def _py_cc_toolchain_impl(ctx):
+    py_cc_toolchain = PyCcToolchainInfo(
+        headers = struct(
+            providers_map = {
+                "CcInfo": ctx.attr.headers[CcInfo],
+                "DefaultInfo": ctx.attr.headers[DefaultInfo],
+            },
+        ),
+        python_version = ctx.attr.python_version,
```
Comment (chrislovecnm) [inline python/private/py_cc_toolchain_rule.bzl:27]: Should we check that the `python_version` is supported?  We often run a check to determine that the version is in the dict:

https://github.com/bazelbuild/rules_python/blob/5b8fa22a2f22501b18b4aea97c5dbfe3a6913a0c/python/versions.bzl#L249

File: python/private/py_cc_toolchain_rule.bzl (hunk 7 above)
Comment (chrislovecnm) [reply python/private/py_cc_toolchain_rule.bzl:33]: Nevermind, the keys are not shorted.  One of the places that buildifier does not force you to sort the keys.

File: python/private/py_cc_toolchain_rule.bzl [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            providers = [CcInfo],
+            mandatory = True,
+        ),
+        "python_version": attr.string(
+            doc = "The Major.minor Python version, e.g. 3.11",
+            mandatory = True,
+        ),
+    },
+    doc = """\
+A toolchain for a Python runtime's C information (e.g. headers)
```
Comment (chrislovecnm) [inline python/private/py_cc_toolchain_rule.bzl:48]: Are we getting more information than just the headers?  Runtime information often goes into the routines and not just the header files.

File: tests/cc/BUILD.bazel [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+)
+
+toolchain(
+    name = "mac_toolchain_definition",
+    target_compatible_with = ["@platforms//os:macos"],
+    toolchain = ":mac_toolchain",
+    toolchain_type = "@bazel_tools//tools/cpp:toolchain_type",
+)
+
+fake_cc_toolchain_config(
```
Comment (chrislovecnm) [inline tests/cc/BUILD.bazel:80]: Tests make me HAPPY!

File: tests/cc/BUILD.bazel [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+)
+
+toolchain(
+    name = "linux_toolchain_definition",
+    target_compatible_with = ["@platforms//os:linux"],
+    toolchain = ":linux_toolchain",
+    toolchain_type = "@bazel_tools//tools/cpp:toolchain_type",
+)
+
+fake_cc_toolchain_config(
```
Comment (chrislovecnm) [inline tests/cc/BUILD.bazel:106]: Can you add a few comments on how the test is working?

File: tests/cc/current_py_cc_headers/current_py_cc_headers_tests.bzl [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        attrs = {
+            "header": attr.label(
+                default = "//tests/cc:fake_header.h",
+                allow_single_file = True,
+            ),
+        },
+    )
+
+def _test_current_toolchain_headers_impl(env, target):
+    compilation_context = env.expect.that_target(target).provider(
```
Comment (chrislovecnm) [inline tests/cc/current_py_cc_headers/current_py_cc_headers_tests.bzl:40]: Can you add a few comments on what and how we are testing?

File: tests/cc/py_cc_toolchain/py_cc_toolchain_tests.bzl [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        meta = env.expect.meta.derive(expr = "py_cc_toolchain_info"),
+    )
+    toolchain.python_version().equals("3.999")
+
+    toolchain_headers = toolchain.headers()
+    toolchain_headers.providers_map().keys().contains_exactly(["CcInfo", "DefaultInfo"])
+
+    cc_info = cc_info_subject(
+        # TODO: Use DictSubject.get once available,
+        # https://github.com/bazelbuild/rules_testing/issues/51
```
Comment (chrislovecnm) [inline tests/cc/py_cc_toolchain/py_cc_toolchain_tests.bzl:52]: Can we file an issue to track this TODO?

File: tests/default_info_subject.bzl [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""DefaultInfo testing subject."""
+
+# TODO: Load this through truth.bzl#subjects when made available
+# https://github.com/bazelbuild/rules_testing/issues/54
```
Comment (chrislovecnm) [inline tests/default_info_subject.bzl:17]: Can we add an issue for this TODO?

File: python/private/py_cc_toolchain_info.bzl [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implementation of PyCcToolchainInfo."""
+
+PyCcToolchainInfo = provider(
```
Comment (chrislovecnm) [inline python/private/py_cc_toolchain_info.bzl:17]: Can we generate the documentation for this API and include the docs in the docs folder?  Are we marking this API GA or Beta?  I prefer Beta until we validate that the API is stable and won't change.

**Review by chrislovecnm — CHANGES_REQUESTED** [review]
Looks great!  The tests are awesome.  Lots of nitpicks.

File: python/cc/BUILD.bazel (hunk 1 above)
Comment (rickeylev) [reply python/cc/BUILD.bazel:5]: Yeah, good idea. Filed https://github.com/bazelbuild/rules_python/issues/1295

File: python/cc/BUILD.bazel (hunk 2 above)
Comment (rickeylev) [reply python/cc/BUILD.bazel:4]: Filed https://github.com/bazelbuild/rules_python/issues/1296

File: tests/cc/py_cc_toolchain/py_cc_toolchain_tests.bzl (hunk 13 above)
Comment (rickeylev) [reply tests/cc/py_cc_toolchain/py_cc_toolchain_tests.bzl:52]: Filed https://github.com/bazelbuild/rules_python/issues/1297

File: tests/default_info_subject.bzl (hunk 14 above)
Comment (rickeylev) [reply tests/default_info_subject.bzl:17]: Filed https://github.com/bazelbuild/rules_python/issues/1297

File: tests/cc/BUILD.bazel (hunk 11 above)
Comment (rickeylev) [reply tests/cc/BUILD.bazel:104]: I think the comment on line 15 through you off -- that was a copy/paste mistake; removed.. There's not tests in this particular BUILD file; this is all just supporting code for the tests you see in the current_py_cc_headers and py_cc_toolchain directories.

File: tests/cc/current_py_cc_headers/current_py_cc_headers_tests.bzl (hunk 12 above)
Comment (rickeylev) [reply tests/cc/current_py_cc_headers/current_py_cc_headers_tests.bzl:41]: Done, I think? The tests here are somewhat hand wavy because all we're doing is forwarding along information from the underlying target. Our fake_headers target just has a single header file, so it doesn't have much within it to check.

I'm not sure what you mean by "how", can you be more specific? These are analysis tests (https://bazel.build/rules/testing#testing-rules) and how they work is fairly well documented by Bazel and rules_testing (a macro setups up a graph of targets, then a rule inspects it to do asserts). The surrounding `env.expect.that_target(...).contains_exactly(...)` stuff are just helper APIs from rules_testing that eliminate a lot of the boiler plate of fetching information and checking state.

File: python/private/py_cc_toolchain_info.bzl (hunk 15 above)
Comment (rickeylev) [reply python/private/py_cc_toolchain_info.bzl:17]: Yeah, definitely beta for now.

I'll see how well these show up as generated docs. I think the fronting macro is going to ruin it and the docs will have to be copy/pasted.

**Review by rickeylev — COMMENTED** [review]
PTAL

File: python/private/py_cc_toolchain_info.bzl (hunk 15 above)
Comment (rickeylev) [reply python/private/py_cc_toolchain_info.bzl:17]: Docs generated

Comment (chrislovecnm) [reply python/private/py_cc_toolchain_info.bzl:17]: Can we mark in the docs that the API is beta and may change?

File: python/private/py_cc_toolchain_rule.bzl (hunk 8 above)
Comment (chrislovecnm) [reply python/private/py_cc_toolchain_rule.bzl:27]: Well, I think we have some other problems with a toolchain that does not match our versions. Are we testing that?

File: tests/cc/current_py_cc_headers/current_py_cc_headers_tests.bzl (hunk 12 above)
Comment (chrislovecnm) [reply tests/cc/current_py_cc_headers/current_py_cc_headers_tests.bzl:41]: Let me rephrase.  The how is it running is with the bazel test rules.  The internals of the test are checking somehow that the new rule is working.  How is the test checking that?  If we changed the code, what do I need to know to update the test?

**Review by chrislovecnm — CHANGES_REQUESTED** [review]
Just a couple of nits

File: python/private/current_py_cc_headers.bzl (hunk 3 above)
Comment (rickeylev) [reply python/private/current_py_cc_headers.bzl:26]: Rephrased to C/C++. CcInfo is for both. It'll provide whatever headers the toolchain provides.

File: tests/cc/current_py_cc_headers/current_py_cc_headers_tests.bzl (hunk 12 above)
Comment (rickeylev) [reply tests/cc/current_py_cc_headers/current_py_cc_headers_tests.bzl:41]: I'm still not really clear about what additional comments you want?

I get the impression that you're not familiar with analysis tests and/or rules_testing's apis, and want comments added describing what analysis tests are, how they work, and how rules_testing's API fetch and assert values? But that's beyond the scope and responsibilities of a single test; we have dozens of analysis tests, see [tools/build_defs/python/tests](https://github.com/bazelbuild/rules_python/tree/main/tools/build_defs/python/tests).

The essence of an analysis test is it's just a rule implementation that looks at the returned providers of a target. What it looks at, how it gets it, how it asserts the values -- that all depends on the particulars of what it's trying to check. So I can't really answer the question of "If we changed the code, what do I need to know?" beyond a vague "you need to know how to get the appropriate value and assert its state".

In this case, all we're doing is getting the CcInfo provider and looking at a couple of its fields.

Comment (chrislovecnm) [reply tests/cc/current_py_cc_headers/current_py_cc_headers_tests.bzl:41]: all good.

**Review by chrislovecnm — APPROVED** [review]
